SalesforceCommerceCloud/pwa-kit
# Merge product bundles feature into develop (@W-16252867@)

# Description

This PR adds support for product bundles into the PWA Kit, and merges the feature branch into the `develop` branch

The code in the feature branch has been reviewed in smaller PRs that were merged into the feature branch. So for reviews, maybe a brief pass through the code for any red flags but mostly looking for manual smoke testing.



# Types of Changes




> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- (change1)

# How to Test-Drive This PR

1. `git checkout feature/product_bundles`
2. `npm ci` in root
3. `npm run start` in `packages/template-retail-react-app`
4. Search for `bundle` (NOTE: for ease of testing, you can reach out to me for details for an environment with a product bundle already configured)
6. Product bundles should work on these pages:

NOTE: If you're using the `default.js` provided in the slack thread, the following variants on the `Women's Business Casual Bundle` have inventory restrictions
- `Swing Tank`: Black, Medium - Out of stock
- `Swing Tank`: Black, Large - Only 3 available
    - selecting a parent bundle quantity of 4 will prompt the inventory message `Only 3 left for Swing tank!`

**PLEASE DO NOT CHECKOUT THESE SPECIFIC VARIANTS FOR TESTING AS IT WILL AFFECT THE INVENTORY** 

- PDP:
    - Selecting `Add bundle to cart` button prompts user to select all choose variants for bundle children first
    - Choosing all variants for bundle children allows user to add bundle to cart
    - User can add bundle to wishlist
    - Choosing out of stock bundle child variant prompts user with inventory message `Out of stock for <product-name>`
    - Choosing quantity that exceeds bundle child variant prompts user with inventory message `Only <X> left for <product-name>`

- Cart Page:
    - Product bundle should appear with parent bundle quantity, variant selection of bundle children, and quantity of bundle children
    - User can increase or decrease quantity using the quantity picker in the product line item
    - If quantity exceeds one of the chosen bundle children variants, user is prompted with inventory message `Only <X> left for <product-name>`
    - User can remove product bundle from cart

- Cart edit modal
    - User can edit the variant selection and parent bundle quantity in edit modal by clicking `Edit` on the cart line item
    - Chosen variant selection is updated and reflecting in the cart line item
    - Choosing out of stock bundle child variant prompts user with inventory message `Out of stock for <product-name>`
    - Choosing quantity that exceeds bundle child variant prompts user with inventory message `Only <X> left for <product-name>`

- Checkout
    - User should be able to see product in order summary dropdown on the side of the checkout page
    - User should be able to checkout product bundle and see product bundles in order confirmation page

- Order history
    - User should be able to see product bundles in order history with quantity and bundle children variant information

- Wishlist
    - User should be able to add bundle to wishlist
    - User should be able to observe bundle in wishlist
        - NOTE: the product line item should only have the bundle children and their quantity and NOT variant selection as the API does not support it currently
    - User should be able to remove bundle from wishlist

# Checklists




## General


## Accessibility Compliance

You must check off all items in **one** of the follow two lists:


_or..._


## Localization
